src/shared/utilities/projectDefaults.ts

```diff
@@ -0,0 +1,87 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as path from 'path'
+import { mkdir, writeFile } from '../filesystem'
+import { fileExists } from '../filesystemUtilities'
+
+export interface SamDeployDefaults {
+    region?: string
+    s3BucketName?: string
+    stackName?: string
+}
+
+export interface AllDafaults {
+    samDeploy: {
+        [samTemplatePath: string]: SamDeployDefaults
+    }
+}
+
+export interface ProjectDefaultsManager {
+    readonly filePath: string
+    refresh(): void
+    getAllDefaults(): AllDafaults
+    save(): Promise<void>
+    setSamDeployDefaults(params: SamDeployDefaults): Promise<void>
+    getSamDeployDefaults(): SamDeployDefaults | undefined
+}
+
+export function makeProjectDefaultsManager({samTemplatePath}: {
+    samTemplatePath: string
+}): ProjectDefaultsManager {
+    let allDefaults: AllDafaults
+    const samProjectDir = path.dirname(samTemplatePath)
+    const filePath = path.resolve(samProjectDir, '.aws-toolkit-vscode/user-prefs.json')
```

Comment:
Can we just use the existing `.aws` directory (and maybe use the current `templates.json` file)?